gcs/src/gcs_gcomm.cpp

```diff
@@ -230,6 +230,18 @@ class GCommConn : public Consumer, public Toplay
 
     ~GCommConn()
     {
+        // We cannot call gcs_close() to perform cleanup if there is an
+        // exception. So, we need to explicity free the objects in the
+        // destructor.
+        if (tp_ != nullptr)
+        {
+            delete tp_;
+        }
+        if (!terminated_)
```

Comment:
Shouldn't terminated_ be tested under the lock Lock lock(mutex_);?
Shouldn't we signal terminated_=true and then join?

Reply:
Fixed this by calling `notify()` after setting `terminated_ = true` with mutex protection.